extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/YoutubeDashManifestCreator.java

```diff
@@ -0,0 +1,1614 @@
+                    "Could not generate or append to the document the SegmentTemplate element of the DASH manifest", e);
+        }
+    }
+
+    /**
+     * Generate the {@code <SegmentTimeline>} element, appended as a child of the
+     * {@code <SegmentTemplate>} element.
+     *
+     * <p>
+     * The {@code <SegmentTemplate>} element needs to be generated before this element with
+     * {@link #generateSegmentTemplateElement(Document, String, DeliveryType)}.
+     * </p>
+     *
+     * @param document the {@link Document} on which the the {@code <SegmentTimeline>} element will
+     *                 be appended
+     * @throws YoutubeDashManifestCreationException if something goes wrong when generating or
+     *                                              appending the {@code <SegmentTimeline>} element
+     *                                              to the document
+     */
+    private static void generateSegmentTimelineElement(@Nonnull final Document document)
+            throws YoutubeDashManifestCreationException {
+        try {
+            final Element segmentTemplateElement = (Element) document.getElementsByTagName(
+                    "SegmentTemplate").item(0);
+            final Element segmentTimelineElement = document.createElement("SegmentTimeline");
+
+            segmentTemplateElement.appendChild(segmentTimelineElement);
+        } catch (final DOMException e) {
+            throw new YoutubeDashManifestCreationException(
+                    "Could not generate or append to the document the SegmentTimeline element of the DASH manifest", e);
+        }
+    }
+
+    /**
+     * Generate segment elements for OTF streams.
+     *
+     * <p>
+     * By parsing by the first media sequence, we know how many durations and repetitions they are
+     * so we just have to loop into {@link #SEGMENTS_DURATION} and {@link #DURATION_REPETITIONS}
+     * to generate for each duration, the following element:
+     * <br>
+     * {@code <S d="segmentDuration" r="durationRepetition" />}
+     * <br>
+     * These elements will be appended as children of the {@code <SegmentTimeline>} element.
+     * </p>
+     * <p>
+     * The {@code <SegmentTimeline>} element needs to be generated before this element with
+     * {@link #generateSegmentTimelineElement(Document)}.
+     * </p>
+     *
+     * @param document the {@link Document} on which the the {@code <S>} elements will be appended
+     * @throws YoutubeDashManifestCreationException if something goes wrong when generating or
+     *                                              appending the {@code <S>} elements to the
+     *                                              document
+     */
+    private static void generateSegmentElementsForOtfStreams(@Nonnull final Document document)
+            throws YoutubeDashManifestCreationException {
+        try {
+            if (isNullOrEmpty(SEGMENTS_DURATION) || isNullOrEmpty(DURATION_REPETITIONS)) {
```

Comment:
Both variables are final List instances. Therefore, they cannot be null.
```suggestion
            if (SEGMENTS_DURATION.isEmpty() || DURATION_REPETITIONS.isEmpty()) {
```